### PR TITLE
Add tests for event-system

### PR DESCRIPTION
**Pull request by PatriceJiang**
Re: #

### Changelog

* add unit-tests for event-system
* use memory pool for list entry of event bus
* rename Listener::event_type to Listener::EventType

-------

### Continuous Integration

This pull request:

* [x] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/core/event/EventBus.h [hunk 1]
```diff
@@ -26,18 +26,29 @@
...
 #include "intl/List.h"
 
 namespace cc {
 namespace event {
-
+class BusEventListenerBase;
+struct BusEventListenerEntry {
+    BusEventListenerEntry *next{nullptr};
+    BusEventListenerEntry *prev{nullptr};
+    BusEventListenerBase *listener = nullptr;
```
Comment (minggo) [inline native/cocos/core/event/EventBus.h]: Please use { } to initialize as others.

File: native/cocos/core/event/EventBus.h [hunk 2]
```diff
@@ -160,6 +165,9 @@ class Listener : public BusEventListenerBase {
 
 template <typename EHandler>
 Listener<EHandler>::Listener() {
+    entry = BusEventListenerEntry::pool()->alloc();
+    entry->next = entry->prev = nullptr;
+    entry->listener = this;
```
Comment (minggo) [inline native/cocos/core/event/EventBus.h:183]: No place to free the memory.

File: native/cocos/core/event/EventBus.h (hunk 1 above)
Comment (PatriceJiang) [reply native/cocos/core/event/EventBus.h]: fixed

File: native/cocos/core/event/EventBus.h (hunk 2 above)
Comment (PatriceJiang) [reply native/cocos/core/event/EventBus.h:183]: fixed

File: native/cocos/core/event/EventBus.h [hunk 3]
```diff
@@ -51,21 +58,15 @@ class BusEventListenerContainer {
...
     }
     void fixPendings();
 
 protected:
-    BusEventListenerBase *_arr{nullptr};
-    BusEventListenerBase *_pendingNew{nullptr};
-    BusEventListenerBase *_pendingDel{nullptr};
+    BusEventListenerEntry *_arr{nullptr};
+    BusEventListenerEntry *_pendingNew{nullptr};
+    ccstd::vector<BusEventListenerEntry *> _pendingDel;
```
Comment (minggo) [inline native/cocos/core/event/EventBus.h]: Please use more meaningful parameter name.

File: native/cocos/core/event/EventBus.cpp [hunk 4]
```diff
@@ -29,33 +29,39 @@
...
+    CC_ASSERT(!listener->entry);
+    listener->entry = new BusEventListenerEntry;
+    listener->entry->listener = listener;
     if (_isBroadcasting) {
-        intl::listAppend(&_pendingNew, lis);
+        intl::listAppend(&_pendingNew, listener->entry);
         return true;
     }
-    return doAddListener(lis);
+    return intl::listAppend(&_arr, listener->entry);
```
Comment (minggo) [inline native/cocos/core/event/EventBus.cpp]: It seems don't have to return a bool value.

File: native/tests/unit-test/src/core_event_bus.cpp [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+int Test1_ListenerT::handleArg3_arg0 = 0;
+char *Test1_ListenerT::handleArg3_arg1 = nullptr;
+float Test1_ListenerT::handleArg3_arg2 = 0;
+
+static int handleArg2_value1 = 0;
+static const char *handleArg2_value2 = nullptr;
+static void handleArg2(int arg0, const char *arg1) {
+    handleArg2_value1 = arg0;
+    handleArg2_value2 = arg1;
+}
```
Comment (minggo) [inline native/tests/unit-test/src/core_event_bus.cpp:42]: It is better to define them in anonymous namespace.

Comment (minggo) [inline native/tests/unit-test/src/core_event_bus.cpp:42]: It is better to define them in anonymous namespace.

File: native/cocos/core/event/EventBus.h (hunk 3 above)
Comment (PatriceJiang) [reply native/cocos/core/event/EventBus.h]: fixed

File: native/cocos/core/event/EventBus.cpp (hunk 4 above)
Comment (PatriceJiang) [reply native/cocos/core/event/EventBus.cpp]: _persist_function_type

File: native/tests/unit-test/src/core_event_bus.cpp (hunk 5 above)
Comment (PatriceJiang) [reply native/tests/unit-test/src/core_event_bus.cpp:42]: fixed

Comment (PatriceJiang) [reply native/tests/unit-test/src/core_event_bus.cpp:42]: fixed

File: native/cocos/core/event/EventBus.cpp [hunk 6]
```diff
@@ -29,33 +29,38 @@
...
+void BusEventListenerContainer::removeListener(BusEventListenerBase *listener) {
     if (_isBroadcasting) {
-        intl::listAppend(&_pendingDel, lis);
-        return true;
+        _listenersToRemove.emplace_back(listener->entry);
+        listener->entry->listener = nullptr;
+        return;
     }
-    return doRemoveListener(lis);
+    bool ret = intl::detachFromList(&_listenerList, listener->entry);
```
Comment (minggo) [inline native/cocos/core/event/EventBus.cpp]: ret is useless

**Review by minggo — APPROVED** [review]

**Pull request merged**